src/main/java/com/google/sps/DataGraph.java

```diff
@@ -309,27 +315,49 @@ public String mutateGraph(Mutation mut) {
    * Modifies the list of tokens of this node to either add or remove tokens contained in tokenMut
    *
    * @param node the node whose token list should be modified
-   * @param tokenMut the mutation that should be applied to the token list
-   * @return the new GraphNode object, or null if it's an unrecognized mutation
+   * @param tokenMut the mutation that should be applied to the token list. The mutation should be
+   *     passed in builder form so that it can be deduplicated if necessary.
+   * @return the new GraphNode object, or null if it's an unrecognized mutation. The method may also
+   *     modify the mutation itself if it performs a duplicate action like adding an existing token.
    */
-  private GraphNode changeNodeToken(GraphNode node, TokenMutation tokenMut) {
+  private GraphNode changeNodeToken(GraphNode node, TokenMutation.Builder tokenMut) {
 
     // List of tokens to add/remove from the existing list
     List<String> tokenNames = tokenMut.getTokenNameList();
     // The existing list of tokens in the node
-    List<String> tokenList = new ArrayList<>();
-    tokenList.addAll(node.tokenList());
+    List<String> existingTokens = node.tokenList();
+    // The modified list of tokens of the node
+    Set<String> tokenList = new HashSet<>();
+    tokenList.addAll(existingTokens);
 
     TokenMutation.Type tokenMutType = tokenMut.getType();
     if (tokenMutType == TokenMutation.Type.ADD_TOKEN) {
       tokenList.addAll(tokenNames);
 
+      if (tokenList.size() != existingTokens.size() + tokenNames.size()) {
+        // Remove tokens that this mutation adds that already exist in the node
+        // we must reinstantiate tokenNames because the one retrieved from
+        // the token mutation is unmodifiable
+        tokenNames = new ArrayList<>(tokenNames);
```

Comment:
is there a reason to set tokenNames to itself/ what exactly is the point of this line?